### ISSUE TITLE
In OJS, a reviewer cannot access his/her recommendations from a previous round.

### ISSUE THREAD

**Issue by pilasou** [pkp/pkp-lib#9453]
As a reviewer reinvited to a new review round, I would like to be able the see the evaluation I provided in previous rounds. 

In OJS, reviewers that are being invited to different review rounds on the same submission cannot access what was their previous recmmandations.

This improvement will offer the opportunity to reviewers to access their previous review reports and thus, better determine how the author made his/her changes to the manuscript. 

The reviewer will be able to access his/her Previous round reports by clicking on a Round button on top of the reviewer process page (as shown below). 
![image](https://github.com/pkp/pkp-lib/assets/19291880/b721acf1-f5ec-4a4b-9be8-d4fff9c64e0a)

When clicking on a cycle button a new window open containing the following information:  

- Recommandation 
- Forms informations 
- Review file sent 
- Cycle dates 
- Article version reviewed during preceding round

![image](https://github.com/pkp/pkp-lib/assets/19291880/cd4b5df2-59d9-4197-bce5-e73aaf053fd9)

### PRs:

Update (2024-04-02)
[pkp/ui-library#313](https://github.com/pkp/ui-library/pull/341) (jardakotesovec repository)
[pkp/pkp-lib#9526](https://github.com/pkp/pkp-lib/pull/9526)
[pkp/ojs#4148](https://github.com/pkp/ojs/pull/4148)

Initial: 
[#9526](https://github.com/pkp/pkp-lib/pull/9526)

**OMP adaptation PRs:**
OJS: pkp/ojs#4258
OMP: pkp/omp#1556
OPS: pkp/ops#668 ***TESTS ONLY***

PKP-LIB: #9920 ***Initial Commits for PKP-LIB included***

### How to test?

1. Make a submission and send it to the Review step.
2. Invite one reviewer.
3. Connect as reviewer and complete the review (fill the form and add a file).
4. As editor, Request revision and select Revisions will be subject to a new round of peer reviews.
5. As an author, submit revision.
6. As editor, open a 2nd round of review.
7. Invite the 1st round reviewer.
8. Connect as the reviewer.
9. On top of the review page, you should see the Round 1 bouton.
10. In the new window, the reviewer can access its answers (forms and files).

**Comment by asmecher** [comment]
PR: https://github.com/pkp/pkp-lib/pull/9526

@Vitaliy-1, could you review this? Thanks!

**Comment by pilasou** [comment]
Hi, @Vitaliy-1, just to let you know, I am a colleage of @nibou230, he will be on vacation for the next few weeks. Back in January. We will follow up when he comes back. Thanks!

**Comment by jardakotesovec** [comment]
@nibou230 Hi, thanks for the contribution, its certainly useful feature. Only drawback is that this is built on legacy stack and we aim to avoid adding new functionality on this old stack, because that means that we would need to migrate it to newer stack at some point anyway.

I would like to ask whether you would have some dev capacity & interest to build this feature on newer stack?

It would be creating frontend part in Vue.js (significant portion of OJS is already on Vue.js) and creating relevant API endpoints that the frontend would use.

If so, we would provide support and guidance how to do it.
In particular
frontend part: @jardakotesovec 
api: @defstat 
UI/UX: @Devika008 

Let us know what you think

Thank you!

Jarda

**Comment by nibou230** [comment]
Hi @jardakotesovec, if our planning let me some spare time, I will do my best to convert this feature on the newer Vue.js stack. Do you have any portion of the application (on Vue.js) that would be similar so I could take some inspiration from it?

Hello @Vitaliy-1, thanks for the comments, if they are still relevant, I will take a look at those once the refactor is done.

**Comment by jardakotesovec** [comment]
@nibou230 That would be great! Some of the things that we are currently using are quite new, therefore there is not enough examples, so I decided to create boilerplate to show how things should come together. New composition API from vue3 might look bit intimidating at first, but once you get use to it, its very expressive way to write logic for components.

Once the tests run, I will merge it to main (probably tomorrow morning), so you can rebase against main and start moving things over.

PR: [ui-library](https://github.com/pkp/ui-library/pull/312), [pkp-lib](https://github.com/pkp/pkp-lib/pull/9622), [ojs](https://github.com/pkp/ojs/pull/4145)

Since I will be merging unfinished work to the main branch, its behind feature flag, to enable displaying of 'History buttons' adjust your config.inc.php to include:
```
[features]
enable_review_round_history = On
```

Now there are two big parts to tackle:

## API
We already have many API endpoints to serve data - https://docs.pkp.sfu.ca/dev/api/ojs/3.4 . Unfortunately the data you are after are not exposed yet. So it will be necessary to introduce new ones that would cover needed metadata. 

Can someone more familiar with that provide bit of guidance and suggest some good modern examples to follow? Pinging @asmecher @Vitaliy-1 @defstat  

In boilerplate for demonstration purposes I am using the /submissions/{submissionId} endpoint to get submission metadata.

## Styling
Thats probably last thing to tackle.
For new UI we are using TailwindCSS - please checkout our [documentation](https://main--6555d3db80418bb1681b8b17.chromatic.com/?path=/docs/guide-style-introduction--docs) in storybook.

Pinging @Devika008 in case she would have capacity to mock what the UI could look like. If not - we will improvise :-).

Also note that one of the parts is listing of reviewer files - I will be building same Vue.js component hopefully in ~2weeks for new submissions listing. So thats something I would provide as well.

Also feel free to explore the storybook that I linked above for additional technical details - I have not promoted it, as I am still working on documentation, but its lots of info already there. More officially it will be introduced in ~2/3weeks.

## Other notes
- Yes I noticed that buttons shows **Round&nbsp;1**, and I am tempted to adjust the translations, but want to check with the team first, rather than fallback to `v-html` which is way to permissive, just to be able pass the html entity.

**Comment by nibou230** [comment]
Wow, thank you so much @jardakotesovec!

I was quite lost trying to figure out how I could start this and how to link the component/page to the reviewer handler section. I'm more of a C# type of developper so it's far from my usual code. :-)

This boilerplate will help me a lot. I'll wait until tomorrow for this code and meanwhile I will look at how I should expose the required data.

Should I create a whole new API controller or should I add a method or two inside the current submission one?

**Comment by defstat** [comment]
Hi @nibou230! Regarding the API, all the data that you are after seem to be contained into the Review process, so it could make sense to start with working on a separate API Controller and add your endpoints there. One example that you could investigate is the [JATS](https://github.com/pkp/pkp-lib/blob/main/api/v1/jats/PKPJatsController.php) API controller. If you have any questions regarding that don't hesitate to ask.

The changes that you intent to develop consider only OJS or OMP also?

**Comment by nibou230** [comment]
Hi @defstat, thank you for the help.

I'll look at the Jats controller and make another one specific for the reviews.

I understand the point to split what should be common or specific to an application, but I don't know OMP at all. I will do my best and if you guys spot something that should not be inside PKP-LIB, tell me and I will make the appropriate change.

**Comment by defstat** [comment]
> Hi @defstat, thank you for the help.
> 
> I'll look at the Jats controller and make another one specific for the reviews.
> 
> I understand the point to split what should be common or specific to an application, but I don't know OMP at all. I will do my best and if you guys spot something that should not be inside PKP-LIB, tell me and I will make the appropriate change.

Perfect, thanks. I would suggest not bothering with OMP specifics if you are not using it. You have already much to consider for this change. :)

**Comment by pilasou** [comment]
HI @Devika008, @defstat, @jardakotesovec Thanks very much for all your help on this! 

Given the extensive refactoring and the fact that we need to adopt a very recent new technology, we propose to make a first MVP of this feature with:  

1. Standard UI using OJS components as is (as documented in [StoryBook](https://main--6555d3db80418bb1681b8b17.chromatic.com/?path=/docs/components-actionpanel--docs)). The page will be structured using the section documented in the issue: Recommendations, Form content, Dates (see image below)
2. For Author Manuscripts and Reviewer file, we will wait for Jarda works on submission to reuse the component he will have created.
3. On first iteration, the form section will only display the default review forms, not custom forms

![rev-history](https://github.com/pkp/pkp-lib/assets/19291880/54023342-7f37-4f46-be2a-0dc8aa5929a7)

Do you agree with this scenario?

**Comment by nibou230** [comment]
Hi guys @defstat @jardakotesovec,

I worked on the different PRs last week:

https://github.com/pkp/ui-library/pull/313
https://github.com/pkp/pkp-lib/pull/9526
https://github.com/pkp/ojs/pull/4148

It's not 100% done, the files are still missing, but you can already take a look if something needs to be addressed right away.

Thanks!

**Comment by defstat** [comment]
@nibou230 hi! Regarding your question on the `reviewerId` and whether [this](https://github.com/pkp/pkp-lib/pull/9526/files#diff-d2395d3a8bc46421de23d6b0d02d0f2af245303a22b2a3a481b9501b68803e19R100) is the way to go, or whether you need to add the `reviewerId` as an input parameter to the request, I think you should consider the specific usecase here - would this issue result to a UI solution that will display "history" information only for the specific reviewer that is currently logged in, or will an editor for example have access to this information also? 

If my understanding is correct, each reviewer will have access to his/her individual history. If this is the case then I think that getting the `reviewerId` from the `request` is sufficient.

**Comment by nibou230** [comment]
OK, thank you! Yes, you are correct, the history side panel is visible only for the current logged in reviewer for now. I wanted to make sure I followed the standards for API development.

**Comment by Devika008** [comment]
Hello @jardakotesovec @nibou230 @pilasou @defstat 

Keeping the elements we currently have and the information architecture we are trying to establish through our revised designs (editorial decisions, email templates, submission dashboard, following is the UX/UI i propose for the issue. 

![image](https://github.com/pkp/pkp-lib/assets/112014397/a992e55c-f701-47d2-bb28-c06aa8283e5a)

### Specifications:
1. I have adapted the [UX/UI of Revised Editorial Decisions](https://github.com/pkp/pkp-lib/issues/7265#issuecomment-1060946918). this includes the stepper mechanism instead of the tabs for a step by step journey. Tabs usually indicate that you can hop between them irrespective of any decision which doesnt support the mental modal here
2. I have also introduced a white block for Previous Reviews. There could be edge cases where there are more than one review round and hence to make the design scalable I introduced the block. 
3. The information flow and design below the stepper remain constant to what is there currently

![image](https://github.com/pkp/pkp-lib/assets/112014397/1e223ee4-3b1a-42e6-b050-144dbcd25374)

### Specifications:
1. On clicking the "Read Review" button, this panel will open just like it does in the summary section of new submission dashboard.
2. The summary section of the dashboard was designed in the way that the grey area was used to support submission information or metadata and the white area to support any information that is important to made a decision. Hence, I've used the grey area here to higlight submission metadata and the general information and the white area to show the recommendations, comments, files and attachments

Let me know if this works for you or if you have any additional comments of feedback on this

**Comment by jardakotesovec** [comment]
@Devika008 Thats great update, thank you! Lets keep this focused on the history. Changing tabs to stepper make sense, but is out of the scope of this one. Maybe create new ticket with that idea and we will see whether we can execute it separately or later when the reviewer process will be revamped.

@nibou230 Based on the design from Devika, here is some guidance that should help you to get that going easily, just merged that to main:

1) 2 Column layout for side modal available in [storybook](https://main--6555d3db80418bb1681b8b17.chromatic.com/?path=/story/components-sidemodal--two-columns-layout) - you will find this example in [SideModal.stories.js](https://github.com/pkp/ui-library/blob/main/src/components/Modal/SideModal.stories.js#L75)
2) Listing files - added component to handle that..  its not pixel perfect per the design, but will refine that later. Here it is in [storybook](https://main--6555d3db80418bb1681b8b17.chromatic.com/?path=/story/listpanel-listingfileslistpanel--base). Source code again in [stories](https://github.com/pkp/ui-library/blob/main/src/components/ListPanel/listingFiles/ListingFilesListPanel.stories.js) file.

Let me know if there is anything else you need. The section with **Read Round** buttons on top should be quite easy just style with Tailwind.

**Comment by Devika008** [comment]
@jardakotesovec Thank you so much :D.

What ill do is that ill create a separate ticket for the stepper for this one. However, I feel it will be a good user experience if we retain the white block above the stepper and use it above the tabs to make it more scalable for the future

**Comment by nibou230** [comment]
Hi everyone, I'm mostly done with the development. You can see a preview of the review history here.

@jardakotesovec The URL used to download the file isn't working. Is this normal (still under development) at this point? It's trying to fetch "http://localhost:8000/index.php/partage/reviewer/submission/url".

![image](https://github.com/pkp/pkp-lib/assets/31410053/a35b39dc-dc4f-43eb-95a2-19b68ede19fa)

**Comment by Devika008** [comment]
This is amazing @nibou230. thank you so much

**Comment by jardakotesovec** [comment]
@nibou230 No idea, so far I just made the component that links to whatever url is passed to it. But to be honest not familiar with the endpoints for downloading files.
I could look at it on Tuesday, but maybe someone more familiar will know the answer.

**Comment by defstat** [comment]
> Hi everyone, I'm mostly done with the development. You can see a preview of the review history here.
> 
> @jardakotesovec The URL used to download the file isn't working. Is this normal (still under development) at this point? It's trying to fetch "http://localhost:8000/index.php/partage/reviewer/submission/url".
> 
> ![image](https://private-user-images.githubusercontent.com/31410053/301637250-a35b39dc-dc4f-43eb-95a2-19b68ede19fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4NTgwMjksIm5iZiI6MTcwNjg1NzcyOSwicGF0aCI6Ii8zMTQxMDA1My8zMDE2MzcyNTAtYTM1YjM5ZGMtZGM0Zi00M2ViLTk1YTItMTliNjhlZGUxOWZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDA3MDg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjYWI5YTA4MmRkZjFiZTgzMjk1MzIyYzhiNDRkMDEzNWEzOTVhMGE3MzkxYjVhYjJkNjQxZTM0ZjE5MTNiMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.03UoG4EqwzSTSIIbcs7Zv-ABRFmhVUa_E1Rt5qwBxK8)

@nibou230 the URL that does not work is the one that you prepared for the needs of this development? If so, you may need to have some code written for the URL to work. We can see that together if this is the case.

**Comment by nibou230** [comment]
Hi @jardakotesovec, I submitted a fix for the URL, you can see it in the last commits. I also had to push another fix for the `SubmissionFileAssignedReviewerAccessPolicy`.

@defstat No, the new component is using the URL fetched with the submissionFile repository. I found the error this morning, the `File` component was trying to set the `href` instead of `:href`.

**Comment by jardakotesovec** [comment]
@nibou230 Thank you for fixing that. Can you please cherry-pick this commit to your ui-library branch to add this page to storybook? https://github.com/jardakotesovec/ui-library/commit/19200fb83992d5ae7559224a8fb0a25cec28dc14  

Would like to try test workflow with Devika, so she can provide feedback in storybook when she can see mostly real thing.

Also you might have better API mocks while testing it - so feel free to replace them with your, especially if it tests wider range of metadata.

**Comment by nibou230** [comment]
@jardakotesovec OK, I already cherry-picked the commit and I will look if I can improve the mocks.

**Comment by jardakotesovec** [comment]
@Vitaliy-1 @defstat @defstat   @asmecher I think last opened problem here is whether it needs some adjustments for OPS/OMP. Do anyone of you know from top of your head what differences there might be for this particular feature? I am not familiar with ops/omp yet much.. will be learning more about it once extending new submission listing for ops&omp.

**Comment by defstat** [comment]
I would suggest, before a more specific review, to add the application specific (OJS) changes to OMP at least, and add PRs for both OMP and OPS for the tests to run. It will give a hint of whether those changes work as they supposed to on OMP and OPS. 

@nibou230 if you have questions on how to prepare the PRs for OMP and OPS (and OJS) so that the tests can run on travis, don't hesitate to ask.

**Comment by pilasou** [comment]
Thanks, @defstat, we can make the PRs for OMP and OPS, but we want to say that we have never work with those softwares. If all goes smoothly and all tests pass for both software, does this mean our PRs will be accepted?

**Comment by nibou230** [comment]
Hi @defstat, the 2 PRs are done :

https://github.com/pkp/ops/pull/647
https://github.com/pkp/omp/pull/1532

**Comment by asmecher** [comment]
Thanks, all! On the OMP and OPS changes, it looks like all the work went into pkp-lib, which is good (and will make it easy to adapt to the other apps). In cases where we have a good contribution for OJS but the contributor doesn't work with the other apps, we usually try to meet them partway. OPS doesn't support peer review, so no further changes will be required there. For OMP, there is a distinction between internal and external review that might need consideration. @defstat, can you give that a look/think?

**Comment by Devika008** [comment]
Hello,

I made some minute changes to the UI and here are the revised designs @jardakotesovec @nibou230.

These include changing the text buttons and the colour of the text from #000000 to #222222
![image](https://github.com/pkp/pkp-lib/assets/112014397/a607f564-afc1-4952-aa03-c650a318506c)

These include changing the text buttons and the colour of the text from #000000 to #222222 and changed the heading colour to 01354F.
Along with that I increased the border size to 1px as pointed out by @jardakotesovec 
![image](https://github.com/pkp/pkp-lib/assets/112014397/71744e11-06f4-4632-b374-eb901356e923)

**Comment by nibou230** [comment]
Hi @Devika008, I followed your recommendations and ajusted the UI. However, the border size is already 1px, so I didn't touch it at all.

At this point, my time remaining on the project is well over and if you wish, I will let you take over if there are any changes needed. Please do not hesitate to contact us if there is anything that requires our assistance.

**Comment by Devika008** [comment]
thank you so much @nibou230 for all the work and adjustments you have done to the interface. Will reach out to you incase of any assistance

**Comment by pilasou** [comment]
Hi, thank all! As @nibou230 mentionned, we are not able allocate more time to this contribution right now. However, from my understanding the feature is functional. Is there any thing remaining to do that prevent our contribution to be merge to main branch?

**Comment by defstat** [comment]
Hi all!

For OMP I think that the only change that we should make is using the `stageId` parameter/filter when ever we are retrieving `reviewRounds` and `reviewAssignments`. By doing that, the `INTERNAL` and `EXTERNAL` stage ids will be properly distinguished and the correct review history will be retrieved from the API call. 

For OMP we should be mindful with `sectionIds` also - we should not consider the submission having one by default. 

I will make some suggested code changes along these lines.

**Comment by jardakotesovec** [comment]
Hi,
I did some styling&html structure updates and resolved conflict with main. I did update the link to the new PR in the first post. I squashed @nibou230 commits together and now just adding some changes on top of it to ensure his authorship stays in place. 

@Devika008 Will be requesting review in chromatic for the updates.

**Comment by jardakotesovec** [comment]
Hi there,
did some additional tweak to also handle scenario when the previous review is confirmed but not really completed. Devika advised to keep it simple and indicate it was declined. As editors often decline it on reviewer behalf if they don't finish it anyway.

@defstat I am happy with frontend part and given that its behind the feature flag I feel like merging this will safe us from some conflict headaches with main branch. 

We can still keep the issue opened until we are confident its all finished, including ops&omp.

**Comment by jardakotesovec** [comment]
@defstat Just merged the ui-library part as I need some of the changes elsewhere..  and since its behind feature flag its not big problem for now that it won't work. But if there is something to adjust on frontend for the omp let me know. Happy to make adjustment

**Comment by defstat** [comment]
Hi everybody. I have added 2 PRs for OMP adaptation. 

I have added the code necessary to distinguish between INTERNAL_REVIEW and EXTERNAL_REVIEW.

I will test OPS on this PKP-LIB code base, expecting that it will not have any problems as the review process is not getting initialized there

**Comment by asmecher** [comment]
@defstat, I took a skim over the PRs adapting this to OMP/OPS, and it looks OK to me -- just one comment. Otherwise, please go ahead.

**Comment by defstat** [comment]
@asmecher thanks. Made the proposed change.

@jardakotesovec there is a `enable_review_round_history` switch in the config.inc.php and in the feature's code. I suppose I shall remove it before merging the feature. That switch was only for the development stage of the issue, correct?

**Comment by jardakotesovec** [comment]
@defstat Yes exactly, this switch should be removed before this issue is closed. Thank you!

**Comment by defstat** [comment]
@jardakotesovec I am waiting for the tests to pass, and then I am going to merge the PRs for OJS, OMP and PKP-LIB. 

Any final thoughts before that?

**Comment by defstat** [comment]
@asmecher @jardakotesovec @Devika008 OJS, OMP, PKP-LIB merged into main.

@pilasou @nibou230 This feature is now part of OJS - Thanks so much for your contribution!

**Comment by pilasou** [comment]
@jardakotesovec @Devika008 @asmecher @defstat  Thank you so much! And special thanks to @nibou230 ! :) That was quite an effort and seeing this feature added to OJS is rewarding to say the least. Encore une fois merci à toutes et tous!
